### PR TITLE
Added the middleware | Adniya

### PR DESCRIPTION
**Pull request by adniyaYousaf**

### CODE REVIEW

File: server/api.js [hunk 1]
```diff
@@ -6,6 +6,43 @@ import { v4 as uuidv4 } from "uuid";
...
+		}
+
+		const sessionResult = await db.query(
+			"SELECT * FROM session WHERE token = $1",
+			[token]
+		);
+		const session = sessionResult.rows[0];
+
+		if (!session) {
+			return res.status(401).json({ code: "INVALID_SESSION" });
```
Comment (soonick) [inline server/api.js]:
```suggestion
			return res.status(401).json({ code: "UNAUTHORIZED" });
```

File: server/api.js [hunk 2]
```diff
@@ -6,6 +6,43 @@ import { v4 as uuidv4 } from "uuid";
...
+		}
+
```
Comment (soonick) [inline server/api.js:32]: You need to also check that the token hasn't expired

File: server/api.js [hunk 3]
```diff
@@ -6,6 +6,43 @@ import { v4 as uuidv4 } from "uuid";
...
+		const userResult = await db.query("SELECT * FROM users WHERE id = $1", [
+			session.user_id,
+		]);
+		const user = userResult.rows[0];
+
+		if (!user) {
+			return res.status(401).json({ code: "USER_NOT_FOUND" });
```
Comment (soonick) [inline server/api.js]:
```suggestion
			return res.status(500).json({ code: "SERVER_ERROR" });
```

If there is a token, but there is no user, it means we did something wrong

File: server/api.js [hunk 4]
```diff
@@ -6,6 +6,43 @@ import { v4 as uuidv4 } from "uuid";
...
+		}
+
+		req.user = user;
+		next();
+	} catch (error) {
+		res.status(500).json({ code: "SERVER_ERROR", message: error.message });
```
Comment (soonick) [inline server/api.js]:
```suggestion
		res.status(500).json({ code: "SERVER_ERROR" });
```

Don't return the error message to the client. It could be a security vulnerability.

**Review by soonick — CHANGES_REQUESTED** [review]

File: server/api.js (hunk 2 above)
Comment (adniyaYousaf) [reply server/api.js:32]: Done

File: server/api.js [hunk 5]
```diff
@@ -6,6 +6,53 @@ import { v4 as uuidv4 } from "uuid";
...
+
+		const currentTime = new Date();
+		if (session.expires_at <= currentTime) {
+			return res.status(401).json({ code: "EXPIRED_SESSION" });
+		}
+
+		const currentTime = new Date();
+		if (session.expires_at <= currentTime) {
+			return res.status(401).json({ code: "EXPIRED_SESSION" });
+		}
```
Comment (soonick) [inline server/api.js]: Why is this here twice?

Comment (soonick) [inline server/api.js]: I think this won't work. Did you test it?

**Review by soonick — CHANGES_REQUESTED** [review]

File: server/api.js (hunk 5 above)
Comment (adniyaYousaf) [reply server/api.js]: Yes. I created a session table in DB with these columns and added dummy data to verify it is working or not.

Comment (adniyaYousaf) [reply server/api.js]: Sorry I merged commits that's why its added twice.

Comment (soonick) [reply server/api.js]: Did you verify that the session will be rejected if the expiration date is older than today?

Comment (adniyaYousaf) [reply server/api.js]: Yes. It is working. I added this dummy data in DB to check it and I got this response  "code": "EXPIRED_SESSION"
INSERT INTO session (user_id, token, expires_at)
VALUES (1, 'abc123def456ghi789jkl067', NOW() - INTERVAL '1 day');
